recipes-core/systemd/systemd/229/0002-nspawn-added-support-for-starting-DHCP-client-in-non.patch

```diff
@@ -0,0 +1,203 @@
+@@ -538,3 +540,87 @@ int veth_extra_parse(char ***l, const char *p) {
+         a = b = NULL;
+         return 0;
+ }
++
++int spawn_dhcp_client(const char *dhcp_client, pid_t *rpid) {
++        pid_t pid;
++        const char *interface = "host0";
++        const char *argv[16];
++
++        pid = fork();
++        if (pid < 0)
++                return log_error_errno(errno, "Failed to fork dhclient: %m");
++        else if (pid == 0) {
++                int nullfd;
++                char *empty_env = NULL;
++
++                nullfd = open("/dev/null", O_RDWR);
++                if (nullfd < 0)
++                        _exit(EXIT_FAILURE);
++
++                if (dup3(nullfd, STDIN_FILENO, 0) < 0)
++                        _exit(EXIT_FAILURE);
++
++                if (dup3(nullfd, STDOUT_FILENO, 0) < 0)
++                        _exit(EXIT_FAILURE);
++
++                if (dup3(nullfd, STDERR_FILENO, 0) < 0)
++                        _exit(EXIT_FAILURE);
++
++                if (nullfd > 2)
++                        safe_close(nullfd);
++
++                (void) reset_all_signal_handlers();
++                (void) reset_signal_mask();
++                close_all_fds(NULL, 0);
++
++                if (dhcp_client) {
++                        char **args = strv_split(dhcp_client, " ");
++                        char **a;
++                        int i;
++
++                        if (args == NULL)
++                                _exit(EXIT_FAILURE);
++
++                        i = 0;
++                        STRV_FOREACH(a, args) {
++                                if (i >= (int)ELEMENTSOF(argv) - 1)
++                                        _exit(EXIT_FAILURE);
++
++                                if (i == 0) {
++                                        char *p = strrchr(*a, '/');
++
++                                        argv[i] = p ? p + 1 : *a;
++                                }
++                                else
++                                        argv[i] = *a;
++                        }
++                        argv[i] = interface;
++
++                        execve(args[0], (char *const *)argv, &empty_env);
++                        _exit(EXIT_FAILURE);
++                }
++                else {
++                        argv[0] = "dhclient";
++                        argv[1] = interface;
++                        argv[2] = NULL;
++
++                        execve("/usr/sbin/dhclient", (char *const*)argv, &empty_env);
++                        execve("/sbin/dhclient", (char *const*)argv, &empty_env);
++
++                        argv[0] = "udhcpc";
++                        argv[1] = "-i";
++                        argv[2] = interface;
++                        argv[3] = NULL;
++
++                        execve("/usr/sbin/udhcpc", (char *const*)argv, &empty_env);
++                        execve("/sbin/udhcpc", (char *const*)argv, &empty_env);
```

Comment:
Yes, I agree that we should use systemd-networkd to configure the interface. We already know it is always available because we bind mount /sbin, /bin, /usr/sbin, and /usr/bin into the runtime container image when we (ask nspawn to) stitch it together. Manually running a DHCP client was a hack to get things rolling and we'll get rid of it eventually.

It is an easy choice what to do when we're running a full-blown system container: we just enable systemd-networkd inside the container and make sure the image has a proper network.conf file for host0. Everything should just work then.

Things get trickier when we run a service-only container. Before systemd v229, such a beast used to be a single-process container with nspawn starting up the target process as PID1 inside the container. This was not a good idea as processes did not handle properly the duties of init (reaping dead children reparented under them by the kernel) which typically lead to a container zombie apocalypse over time, if you were running a forking server inside the container. Hence, starting from v229 one can (and should practically always) use the --as-pid2 option to turn such a container into a dual-process setup, with the target process being run as PID2 and PID1 being a stub init which takes care of properly reaping children processes.

Now with this change and such a setup, I think the best solution would be to make the stub init a bit smarter so that it could take care of properly configuring the container network interfaces as well (either by itself or by starting up systemd-networkd in addition to the target process).

A third really brute-force option is to stop supporting anything else but full-blown system containers. But I'd rather not make such a brutal decision, at least not yet. Let's see how difficult it will be to support such a setup in the end, and make that call if the pain/gain ratio is getting too big.